packages/beacon-state-transition/src/cache/epochContext.ts

```diff
@@ -400,6 +417,7 @@ export class EpochContext {
     const nextEpoch = currEpoch + 1;
     this.nextShuffling = computeEpochShuffling(state, epochProcess.nextEpochShufflingActiveValidatorIndices, nextEpoch);
     this.proposers = computeProposers(state, this.currentShuffling, this.effectiveBalanceIncrements);
+    this.nextEpochProposers = computeProposers(state, this.nextShuffling, this.effectiveBalanceIncrements);
```

Comment:
I think we should only calculate this on demand (when the api call) and cache it, instead of on every epoch transition

Reply:
I actually thought about that also, but decided to go with this implementation due to its simplicity. (No need for code to only generate this after the first call, and then code to purge previous values when a new epoch is reached) It also does not introduce any performance regression, that is why I went with it.

Reply:
> I actually thought about that also, but decided to go with this implementation due to its simplicity

Should not be complicated at all to implement. Just type as T | null and compute if requested and null.

> It also does not introduce any performance regression, that is why I went with it.

Our performance CI only rejects x3 fn run time changes. computeProposers() has a cost of ~100ms over the state transition which takes 1000ms, so a ballpark 10% increase aprox.

Reply:
> Should not be complicated at all to implement. Just type as T | null and compute if requested and null.

The complexity is not about the typing though. It's more about passing `CachedBeaconStateAllForks` on the http requests to `computeProposers` - something which is already handled in the transition phase. Also adding a prune call to remove past caches - something which is not needed with this approach as current and next proposers are always refreshed on state transition.

> Our performance CI only rejects x3 fn run time changes. computeProposers() has a cost of ~100ms over the state transition which takes 1000ms, so a ballpark 10% increase aprox.

Okay. Avoiding a ~10% performance hit not a bad thing. I'll update to have it computed on demand.

Reply:
Together with @tuyennhv we looked at how to get the state needed by `computeProposers` which makes implementing the suggestion to compute on demand possible as can be seen in this commit [here](https://github.com/ChainSafe/lodestar/pull/3782/commits/7d08adaf1f98ec5a9151122477b9e816956cc908). 

A couple of observation where I think this approach makes the code less straightforward:

- `getNextEpochBeaconProposer` has to be moved to `BeaconStateContext` in other to have access to the `allForks.BeaconState` needed for `computeProposers` (while the other similar method `getBeaconProposer` remains on `EpochContext`)
- A `getState` has to be added to `BeaconStateContext` to get value of `allForks.BeaconState`
- Even though `getNextEpochBeaconProposer` is on `BeaconStateContext` it still has to reach to `EpochContext` to update the cache
-`getBeaconProposer` method now need to have knowledge of the previous proposer cache since there is now the need to clear the previous `nextEpochProposers` in the cache on request for current proposer.

I guess at the end of the day it boils down to whether this is okay and better than incurring the ~10% decrease in performace - which I am fine with.

Also open to other suggestions that could simply the compute at demand approach.